### PR TITLE
Add support for zipped dags

### PR DESCRIPTION
**Pull request by bolkedebruin**
Currently dags are being read directly from the filesystem. Any
hierarchy (python namespaces, modules) need to be reflected on
the filesystem. This makes it hard to manage dags and their
depedencies.

This patch adds support for dags in zip files. It will add
the zip to sys.path and then it will read the zip file and
try to import any files as modules that are in the root of
the zip.

Please note that any module contained within the zip will
overwrite existing modules in the same namespace.

### CODE REVIEW

**Comment by bolkedebruin** [comment]
@mistercrunch Can you provide some help maybe on why travis is failing? At first it seems unrelated to my code, but it only fails in my build so it obviously is related. It's failing on deepcopy so that's why I am asking you.

@criccomini for testing

**Comment by bolkedebruin** [comment]
@mistercrunch Fixed the deepcopy issue with this PR

File: airflow/models.py [hunk 1]
```diff
@@ -198,53 +200,93 @@ def get_dag(self, dag_id):
...
+                    self.file_last_changed[filepath] = dttm
+
+        else:
+            zip_file = zipfile.ZipFile(filepath)
+            for mod in zip_file.infolist():
+                head, tail = os.path.split(mod.filename)
+                mod_name, ext = os.path.splitext(mod.filename)
+                if not head and (ext == '.py' or ext == '.pyc'):
+                    if mod_name == '__init__':
+                        self.logger.warning("Found __init_.py at root of ".format(filepath))
```
Comment (jlowin) [inline airflow/models.py]: `"Found __init__.py at root of {}"` (add one underscore and curly braces)

File: docs/tutorial.rst [hunk 2]
```diff
@@ -411,6 +411,32 @@ which are used to populate the run schedule with task instances from this dag.
     # start your backfill on a date range
     airflow backfill tutorial -s 2015-06-01 -e 2015-06-07
 
+Packaging dags
+''''''''''''''
+While often you will specify dags in a single ``.py`` file it might sometimes
```
Comment (criccomini) [inline docs/tutorial.rst]: Could you add some docs about how to install dependencies into a ZIP?

**Comment by criccomini** [comment]
I'm +1 on this, but I think @mistercrunch / @r39132 might want to take a look.

**Comment by mistercrunch** [comment]
I'm not big on this PR, missed it earlier this week, especially the `sys.path` handling could lead to almost impossible to debug problems. 

I'm not sure why it's necessary. How is a zip file better than a python package?

**Comment by bolkedebruin** [comment]
@mistercrunch The main reasons for having this in are:
1. In the case of not being able to install packages on the OS, ie. in our environment there is a separation of concerns so we are not allowed to install packages or update them, you can now package your DAGs together with dependencies.
2. Versioning: Although this is a bit in the future, we could allow airflow only to pickup dependencies from site-wide or from a "dag-zip" to allow multiple versions.
3. Combining multiple dags into 1 zip for better manageability.

Where our perspectives might differ is that we are not using "git sync" and we have the need to make the addition of dags authenticated and easier to manage. This is first step towards that.

Adding it to sys.path is standard python and is fully supported, but to address your concerns I could make it configurable by adding a "support_zip_dag" and "support_zip_dependencies" and a a log.info when a dependency is picked up and added to the sys.path. Would that be better?

**Comment by mistercrunch** [comment]
The team here wants to revert this as it seems specific to your environment. 

1.) we're planning on  solving with docker support, 2) we have a proposal around versioning coming up, 3) it's all in a repo anyways

Is there a way that it could live in `contrib` or something have some sort of interface?

**Comment by bolkedebruin** [comment]
I need to think about that a little bit on how to approach that and put it in contrib. If you have a suggestion that would be nice.

Could you elaborate about the (remaining) concerns as it was reviewed by @criccomini and @jlowin ? As it stands if you don't use ZIPs nothing will change, if I make it configurable the code won't be evaluated. I could refactor the code and split the function and make it more clear what code paths are used if that helps, but there seems to be an underlying issue that I can't make out.

**Comment by criccomini** [comment]
> it seems specific to your environment.

@mistercrunch, I actually think that deploying DAGs via git is a bit specific to your environment. I know it's how FB does it with dataswarm, but this is a really uncommon practice across a wide range of systems. If you look at virtually every other scheduler, as well as most other distributed deployment systems (mesos, YARN, Kubernetes, etc), none of them deploy code this way.

I'm in a similar position to Bolke--installing a bunch of teams' packages in either the Airflow virtual environment, or the system's Python space is not easily do-able.

What are the exact concerns? I see two, so far:
1. The sys.path issue.
2. The Docker vs. Zip issue.

(1), I'm not well versed to comment on, so I'll leave you two to it. :) For (2), even if Docker DAG packages are supported, for those not using Docker, one could argue that supporting zip deployables might be nice. It's kind of a kitchen sink approach, though, so I'm more in favor of an either/or (I prefer to have the framework be opinionated).

Maybe you could shed some light on what the Docker proposal looks like, when it might be implemented, etc?

**Comment by ldct** [comment]
@bolkedebruin I believe the changes to the `/code` endpoint in `views.py` are incorrect, it will fail for dags with the same `__name__`. See https://issues.apache.org/jira/browse/AIRFLOW-443

**Comment by emillynge** [comment]
I'm sorry for commenting on such an old (and closed) PR.

Is this feature completely implemented?  It seems that the SchedulerJob class uses a completely different method for loading DAGs than the webserver etc. and it will not schedule tasks from a packaged dag since it only considers raw python files.

**Comment by bolkedebruin** [comment]
I need to have a look at that. @aoen @saguziel care to comment? These changes to the scheduler were by @plypaul

**Pull request merged**